File: Obchisluvannya/threaddemosharp/threaddemo/ThreadStopper.cs
using threaddemo;

public class ThreadStopper
{
    private readonly Thread thread;

    private static Dictionary<int, int> timers = [];
    private static List<ThreadCalculator> threads = [];

    public ThreadStopper()
    {
        thread = new Thread(new ParameterizedThreadStart(method => Stopper()));
    }

    private static void Stopper()
    {
        var timersSorted = timers.OrderBy(x => x.Value).ToList();

        int time = 0;
        foreach (var timer in timersSorted)
        {
            Thread.Sleep(timer.Value - time);
            time = timer.Value;

            threads[timer.Key - 1].Stop();
        }
    }
    public void AddThread(ThreadCalculator thread)
    {
        threads.Add(thread);
        var random = new Random().Next(2000, 10000);
        timers[thread.ThreadId] = random;
        System.Console.WriteLine(random);
    }

    public void Start()
    {
        thread.Start();
    }
}

File: Obchisluvannya/threaddemo/src/com/company/ThreadStopper.java
package com.company;

import java.util.*;
import java.util.stream.Collectors;

public class ThreadStopper {
    private final Thread monitorThread;
    private final Map<Integer, Integer> timers = new HashMap<>();
    private final List<ThreadCalculator> threads = new ArrayList<>();

    public ThreadStopper() {
        this.monitorThread = new Thread(this::stopperLogic);
    }

    private void stopperLogic() {
        // Сортуємо таймери за значенням (часом очікування)
        List<Map.Entry<Integer, Integer>> sortedTimers = timers.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByValue())
                .collect(Collectors.toList());

        int elapsedTime = 0;
        for (Map.Entry<Integer, Integer> entry : sortedTimers) {
            try {
                int waitTime = entry.getValue() - elapsedTime;
                if (waitTime > 0) {
                    Thread.sleep(waitTime);
                }
                elapsedTime = entry.getValue();

                // Зупиняємо відповідний потік (id - 1, бо в списку індексація з 0)
                threads.get(entry.getKey() - 1).stop();
                
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    public void addThread(ThreadCalculator thread) {
        threads.add(thread);
        int randomTime = (int) (Math.random() * (10000 - 2000) + 2000); // 2000 - 10000 ms
        timers.put(thread.getThreadId(), randomTime);
        System.out.println("Thread " + thread.getThreadId() + " will stop in: " + randomTime + " ms");
    }

    public void start() {
        monitorThread.start();
    }
}